keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,992 @@
+        are deleted, the finalizer is removed.
+
+#### Example: Allocations
+
+```go
+[
+  {
+      // For existing clusters this is the same as ClusterCIDR
+      CIDR:            "10.0.0.0/8",
+      // Default for nodes not matching any other rule
+      NodeSelector:    nil,
+      // For existing API this is the same as NodeCIDRMaskSize
+      PerNodeMaskSize: 24,
+  },
+  {
+      CIDR:            "172.16.0.0/14",
+      // Another range, also allocate-able to any node
+      NodeSelector:    nil,
+      PerNodeMaskSize: 24,
+  },
+  {
+      CIDR:            "10.0.0.0/8",
+      NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+      PerNodeMaskSize: 26,
+  },
+  {
+      CIDR:            "192.168.0.0/16",
+      NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+      PerNodeMaskSize: 26,
+  },
+  {
+      CIDR:            "5.2.0.0/16",
+      NodeSelector:    { "np": "np3" },
+      PerNodeMaskSize: 20,
+  },
+  ...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{"np": "np1"} --> "10.0.0.0/26"
+{"np": "np2"} --> "192.16.0.0/26"
+{"np": "np3"} --> "5.2.0.0/20"
+{"np": "np4"} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new
+[NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam)
+The controller will set up watchers on the `ClusterCIDRConfig` objects and the
+`Node` objects.
+
+This controller relies on being a single writer (just as the current NodeIPAM
+controller does as well). In the case of HA control planes with multiple
+replicas, there will have to be some form of leader election to enforce only 1
+active leader. This KEP proposes re-using the kube-controller-manager leader
+election to pick a active controller.
+
+#### Data Structures
+
+We will use maps to store the allocated ranges and which node is using the
+range. Because the number of nodes is expected to be on the order of thousands,
+more sophisticated data structures are likely not required.
+
+Prior investgations [here](https://github.com/kubernetes/kubernetes/pull/90184)
+suggest that maps storing allocations will perform well under the number of
+nodes we expect.
+
+#### Dual-Stack Support
+
+Before assigning any Pod IPs to nodes, the controller will check all Nodes for
+the IP families in use. If all Nodes are consistent (all single-stack or all
+dual-stack), then the controller enters that mode.
+
+If there are no Nodes or all their `PodCIDR` allocations are empty, the
+controller chekcks and the list of `ClusterCIDRConfigs` for their IPFamilies. If
+there are both IPv4 and IPv6 Pod CIDRs specified, it enters dual-stack mode,
+otherwise it uses single-stack mode.
+
+The controller will wait until the first Node is added before picking a mode. So
+users who want to specify dual-stack must first add 2 `CluterCIDRConfigs` (one
+each for IPv4 and IPv6). Only after creating those resources should Nodes be added.
```

Comment:
just to confirm, migrating a node from single-stack to dual-stack is not possible, you have single-stack nodes, you have to create new nodes, right?

Reply:
That's correct. What I'm unsure of is whether we want to support migrating the cluster from single-stack to dual-stack while nodes are running.

For example, the model proposed here says that if any Node is in single-stack mode, the controller will only allocate single-stack CIDRs to new nodes even if there are IPv4 and IPv6 ClusterCIDRConfigs provided. The only way to migrate your cluster is to add both IPv4 and IPv6 ClusterCIDRConfigs and then delete all nodes. Then once you start adding nodes again, the controller will allocate them in dual-stack mode. Is this an acceptable workflow for us -- I have seen concern about mixing single-stack and dual-stack nodes hence this setup.